### PR TITLE
Docker-ized app setup

### PR DESCRIPTION
**Pull request by ibeckermayer**
Relatively large PR to nail down some of the core app structure, manage self-signed SSL, and run a Docker-ized development and production basic app.

To test:
- See the updated `README.md` for instructions for setting up the dev/prod environments. If anything is confusing or not adequately explained, the `README.md` should be updated.
- When you navigate to the app in the browser, you should see the login form:
  - Invalid email string (i.e. no "@" character), or empty email or password should cause the html5 form validator to complain (screenshot 1)
  - Invalid credentials should cause a 401 error message to display in the console (screenshot 2)
  - Valid credentials (username and password both `admin@goteleport.com`) should take you to the dashboard route (screenshot 3)
  - (Behavior should be exactly the same for both dev and prod setups)

Screen Shot 2021-01-05 at 12 17 06 PM
Screen Shot 2021-01-05 at 12 19 00 PM
Screen Shot 2021-01-05 at 12 21 17 PM

### CODE REVIEW

File: backend/Dockerfile
Comment (russjones) [inline backend/Dockerfile]: Why not just build the application and start it?

File: backend/main.go
Comment (russjones) [inline backend/main.go]: I wouldn't include this logic in `main`. Instead create a package that has your server in it and start that instead. Ideally your `main` would just be flag parsing and then calling that package to start the server. Maybe doing some signal handling but I think that's out of scope for your task.

File: backend/internal/handlers/login.go
Comment (russjones) [inline backend/internal/handlers/login.go]: If the login implementation is not complete, I would not include it. It's hard to review incomplete code.

File: backend/internal/handlers/spahandler.go
Comment (russjones) [inline backend/internal/handlers/spahandler.go:49]: Shouldn't you just return a 404 and only `index.html` if it's requested? @alex-kovoy thoughts?

Comment (ibeckermayer) [reply backend/internal/handlers/spahandler.go:49]: My understanding is that it's written that way in order to support browser-routing (in this case, `react-router`). With this configuration, if a user opens a new tab and navigates directly to a browser-routed route like https://localhost:8000/dashboard, the server will fall back to the index.html and just serve them the react app, which will then recognize the route and load the Dashboard component.

A downside to this is that if the user navigates directly to https://localhost:8000/dne, they'll be shown a blank page. A way that I've dealt with that issue in the past is by creating a "404 not found" react component and route and then redirecting any non-existent routes to it (though the server would in that case still return index.html/200).

File: backend/internal/handlers/login.go
Comment (ibeckermayer) [reply backend/internal/handlers/login.go]: Hmm, yeah makes sense that it's hard to review incomplete code. My intention with this half-baked endpoint was to give me something to hit to make sure that all the other moving parts (SSL/Docker/serving-a-spa-with-go/browser-side-routing) were actually configured correctly.

My initial thinking was that adding the database and session management would be my next PR, though I could just tack that on to this one if you think that makes more sense. Alternately I could make it clear that this is just a toy by renaming it something like "fake_endpoint". Open to better ideas.

File: backend/main.go
Comment (ibeckermayer) [reply backend/main.go]: Gotcha, will refactor this into something cleaner/extensible.

File: backend/Dockerfile
Comment (ibeckermayer) [reply backend/Dockerfile]: What's happening here is that I run the app with `CompileDaemon`, and then in `docker-compose-dev.yml` I mount the `backend/` directory onto this container. `CompileDaemon` then detects whenever I make any changes to any `backend/*.go` files and recompiles/serves the app, effectively giving me a hot-reload server for development. 

I think the confusion may stem from the fact that this Dockerfile is only used for the development setup? If that's the case then I could clarify by renaming it `dev.Dockerfile`

File: backend/internal/handlers/spahandler.go
Comment (alex-kovoy) [reply backend/internal/handlers/spahandler.go:49]: @ibeckermayer you are right, the server needs to return either requested file (if found) or index.html. Client side will handle the rest by showing `404` error component. This should be OK for this project.

File: backend/internal/handlers/spahandler.go
Comment (alex-kovoy) [inline backend/internal/handlers/spahandler.go]: It's better not to leak any information about server file system via error messages. In this particular case it would be enough to return a generic error message ("Invalid request" for example) and log the actual error message.

File: backend/internal/handlers/login.go
Comment (alex-kovoy) [inline backend/internal/handlers/login.go]: Here we are allowing an attacker to know more about our system by leaking this error text. A generic "Access Denied" would suffice (+ log the actual error of course)

File: backend/internal/handlers/utils.go
Comment (alex-kovoy) [inline backend/internal/handlers/utils.go]: i like the max size check

File: backend/internal/handlers/utils.go
Comment (alex-kovoy) [inline backend/internal/handlers/utils.go]: This error message handling seems redundant, just wonder if the Decode error text is too encryptic that it requires additional handling.

Hm...would this be enough? 

```golang
if err != nil {
   return &malformedRequest{status: http.StatusBadRequest, msg: err}
}
```

File: frontend/src/index.jsx
Comment (alex-kovoy) [inline frontend/src/index.jsx:12]: I would actually keep the index route for the dashboard and move the Login to `/login` route.

Also, consider adding Authenticated (or similar) to secure private routes and handle redirects to the login screen when access token is invalid.

```tsx
      <Switch>
        <Route exact path="/">
          <Login />
        </Route>
        <Route path="/">
           <Authenticated>
              <Dashboard />
            <Authenticated/>
        </Route>
      </Switch>
```

File: frontend/src/components/Login/Login.jsx
Comment (alex-kovoy) [inline frontend/src/components/Login/Login.jsx]: kudos for trying to use native APIs (fetch). The only issue with it is that 
> the Promise returned from fetch() won’t reject on HTTP error status even if the response is an HTTP 404 or 500. Instead, it will resolve normally (with ok status set to false), and it will only reject on network failure or if anything prevented the request from completing.

You can add [missing error handling yourself](https://github.com/github/fetch#handling-http-error-statuses) or use axios or similar library instead.

Comment (alex-kovoy) [reply frontend/src/components/Login/Login.jsx]: also consider creating some wrappers for get|post requests to incapsulate and reuse the code that initializes the HTTP request. 

```ts
api.get("/xxx")
api.post('/xxx', json)
```

File: frontend/package.json
Comment (alex-kovoy) [inline frontend/package.json:42]: kudos for not using Create React App boilerplate :) (there is nothing wrong with it aside the fact that it uses a lot of dependencies this is why I also prefer a minimalistic custom setup when possible)

File: backend/internal/handlers/utils.go
Comment (ibeckermayer) [reply backend/internal/handlers/utils.go]: Valid question, the author's justification in the [linked blog](https://www.alexedwards.net/blog/how-to-properly-parse-a-json-request-body) is

> The error messages returned by Decode() aren't ideal for sending to a client. Some are arguably too detailed and expose information about the underlying program (like "json: cannot unmarshal number into Go struct field Person.Name of type string"). Others aren't descriptive enough (like "unexpected EOF") and some are just plain confusing (like "invalid character 'A' looking for beginning of object key string"). There also isn't consistency in the formatting or language used.

which makes sense to me. I wouldn't go through the trouble of writing this all out myself for this project, but since this guy has already done most of the heavy lifting and published it for free, I figure I may as well take the help.

File: backend/internal/handlers/login.go
Comment (russjones) [reply backend/internal/handlers/login.go]: That's fine, will review it in the next PR.

File: backend/internal/handlers/spahandler.go
Comment (russjones) [reply backend/internal/handlers/spahandler.go:49]: @alex-kovoy @ibeckermayer I see, thanks for the explanation.

File: backend/Dockerfile
Comment (russjones) [reply backend/Dockerfile]: That's fine, makes sense now.

File: frontend/package.json
Comment (ibeckermayer) [reply frontend/package.json:42]: Yeah, same here

File: backend/internal/handlers/login.go
Comment (ibeckermayer) [reply backend/internal/handlers/login.go]: Good point. I suppose since all of these endpoints are meant to be private, I should keep responses as generic as possible (give or take what I need to know for the client). i.e in this case, this becomes
```go
	if err != nil {
		var mr *malformedRequest
		if errors.As(err, &mr) {
			log.Println(mr.Error())
			http.Error(w, http.StatusText(mr.status), mr.status)
```

File: backend/internal/handlers/utils.go
Comment (alex-kovoy) [reply backend/internal/handlers/utils.go]: I would say that a client (in this case web ui) does not need to know these details, some generic message "Unable to process request" would suffice and we can log detailed Decode error message as is on the server side. 

If API itself was a product (as `stripe` for example) then I would of course give more attention to error messages returning to the client, but in this particular case it is not worth it. 

For example, would we want to show a user this message in UI?
```
Request body contains an invalid value for the "xxx" field (at position 454)
```

Yeah, lets remove this for now (i do agree that Decode messages might be hard to read but I would not try to optimize it). 

One thing that I do not like is that the author logic will break if internals of the Decode method change, like this string comparsion `strings.HasPrefix(err.Error(), "json: unknown field ")`

File: README.md
Comment (awly) [inline README.md:14]: could you add a script to install mkcert and run this command (ideally dockerized), to reduce the steps needed to try this out?

File: backend/internal/handlers/utils.go
Comment (awly) [inline backend/internal/handlers/utils.go]: you're vendoring the entire godoc.org implementation for a relatively simple function.
you can replace it with `if !strings.HasPrefix(r.Header.Get("Content-Type", "application/json"))` and drop a large dependency

File: backend/internal/handlers/login.go
Comment (awly) [inline backend/internal/handlers/login.go]: nit: extract this into a function, so that the fake authentication logic is contained and separate from normal code

File: backend/internal/handlers/spahandler.go
Comment (awly) [reply backend/internal/handlers/spahandler.go:49]: This seems to create a tight coupling between frontend and backend logic.
But I don't really write web applications these days, so maybe I just need to get on with the times.

File: backend/internal/handlers/utils.go
Comment (awly) [inline backend/internal/handlers/utils.go]: nit: you don't need to use `malformedRequest` as a pointer everywhere

File: backend/internal/handlers/utils.go
Comment (awly) [inline backend/internal/handlers/utils.go]:
```suggestion
	r.Body = http.MaxBytesReader(w, r.Body, 1048576) // 1MB
```

File: backend/main.go
Comment (awly) [inline backend/main.go]: key/cert paths should be passed as flags (with reasonable defaults)

File: backend/internal/handlers/utils.go
Comment (ibeckermayer) [reply backend/internal/handlers/utils.go]: I dug into this a bit and it looks like there's a good reason for using pointer receivers for custom error types

https://stackoverflow.com/a/50333850

Comment (awly) [reply backend/internal/handlers/utils.go]: IMO `malformedRequest{status: 400, msg: "foo"} == malformedRequest{status: 400, msg: "foo"}` is a good property to have.
The linked post doesn't really explain why it's a bad thing, even in `io.EOF` case.

File: frontend/src/index.jsx
Comment (ibeckermayer) [reply frontend/src/index.jsx:12]: I have a plan in my head for how to implement this, but I think it makes more sense to do it when I add session management on the server as well. Ok if I punt it to my next PR?

File: README.md
Comment (ibeckermayer) [reply README.md:14]: I think this could get tricky, since when you run `mkcert -install` the program needs to access your system trust store and add a CA cert/key. Possibly this could work in a Docker container, but then I think you would need to run the browser from that container in order for it to trust the CA.

I could experiment with it but frankly seems more trouble than its worth for this project.

File: backend/internal/handlers/utils.go
Comment (ibeckermayer) [reply backend/internal/handlers/utils.go]: Hmm yeah I suppose I hastily assumed that random guy on the internet had a good point because he wrote a lot of detail, but I also don't see a reason why and agree that the `io.EOF` case doesn't cut it. Will change.

**Review by alex-kovoy — APPROVED** [review]
For the first PR looks good to me, please wait for another approval before merging.

File: backend/internal/server/server.go
Comment (russjones) [inline backend/internal/server/server.go]: You don't want to `log.Fatal` here. Return an error here, you can `log.Fatal` in `main` if you want.

File: backend/internal/server/server.go
Comment (russjones) [inline backend/internal/server/server.go]: Minor issue, but I think you can just do `log.Printf`.

File: backend/main.go
Comment (russjones) [inline backend/main.go:17]: This is not an issue and you don't need to make this change, but a nice pattern we use is the following:

```go
srv, err := server.New(&server.Config{
  Port:        *port,
  Certificate: *certFilePath,
  Key:         *keyFilePath,
})
```

File: backend/main.go
Comment (russjones) [inline backend/main.go:13]: I don't think this works, if I am not mistaken you have to call `flag.Parse()` before using the variables.

**Review by russjones — APPROVED** [review]

File: backend/main.go
Comment (ibeckermayer) [reply backend/main.go:17]: Roger that, will keep that in my back pocket assuming the rest of the interview process goes well :)

**Pull request merged**
